docs/design/endpoints.md

```diff
@@ -0,0 +1,292 @@
+         */
+    }
+}
+```
+
+## Example
+
+The following is an example ruleset attached to an arbitrary service definition with the ID "SmithyService" (the
+definition itself is omitted).
+
+The ruleset describes a simple if-else case: if the sole required parameter `ResourceId` contains a specific prefix,
+route this request against a special-case endpoint. Otherwise, fall back to a constant default. The `substring` function
+used in the ruleset is an example of a standard-library function used by endpoint resolvers.
+
+This particular ruleset is kept simple for the purpose of example, naturally, a ruleset can be of arbitrary complexity as
+required by the service.
+
+```json
+{
+  "parameters": {
+    "ResourceId": {
+      "type": "string",
+      "required": "true"
+    }
+  },
+  "rules": [
+    {
+      "documentation": "use a special endpoint for government resources",
+      "type": "endpoint",
+      "conditions": [
+        {
+          "fn": "substring",
+          "argv": [
+            {"ref": "ResourceId"},
+            0,
+            5,
+            false
+          ],
+          "assign": "resourceIdPrefix"
+        },
+        {
+          "fn": "stringEquals",
+          "argv": [
+            {"ref": "resourceIdPrefix"},
+            "gov."
+          ]
+        }
+      ],
+      "endpoint": {
+        "url": "https://gov.api"
+      }
+    },
+    {
+      "documentation": "fallback to global endpoint",
+      "conditions": [],
+      "endpoint": {
+        "url": "https://global.api"
+      }
+    }
+  ]
+}
+```
+*Note*: The signature of the substring method is `substring(s: String, start: Int, end: Int, reverseIndices: false): String`.
+The argv property in the corresponding function call describes how to source the arguments (or input constants directly
+in the case of the indices).
+
+Given a sample operation `GetResource`, with a single operation input `resource` bound to the `ResourceId` provider
+parameter via context traits, this ruleset would roughly generate the following code:
+
+```kotlin
+public data class SmithyServiceEndpointParams(
+    public val resourceId: String?,
+) {
+    init {
+        require(resourceId != null) { "param resourceId is required" }
+    }
+}
+
+public fun interface SmithyServiceEndpointProvider {
+    open suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint
+}
+
+public class DefaultSmithyServiceEndpointProvider: SmithyServiceEndpointProvider {
+    public suspend fun resolveEndpoint(params: SmithyServiceEndpointParams): Endpoint {
+        val resourceIdPrefix = substring(params.resourceId, 0, 5, false)
+        if (resourceIdPrefix == "gov.") {
+            // use a special endpoint for government resources
+            return Endpoint(
+                Url.parse("https://gov.api"),
+            )
+        }
+
+        // fallback to global endpoint        
+        return Endpoint(
+            Url.parse("https://global.api"),
+        )
```

Comment:
or something like:

```kotlin
val url = when (resourceIdPrefix) {
     "gov." -> "https://gov.api"
     else -> "https://global.api"
}
return Endpoint(Url.parse(url))
```

This is of course an implementation detail we can go over later when we actually have the full scope of rules to deal with.

Reply:
Yeah, I had thought about this, unfortunately we don't really have a good way to "label" the symbols - it would just have to be
```kotlin
// documentation field for 1st
object Endpoint0 : Endpoint ( ... )
// documentation field for 2nd
object Endpoint1 : Endpoint ( ...)
// etc.
```

I think exposing them publicly is IDEAL, though - especially since if people write their own resolvers, they might still want to use some of those endpoints.

I think it's worth asking if we can have the spec label them some way such that we can give them meaningful variable names.

Reply:
I disagree on exposing them unless that is something every SDK is expected to do. Anything we make public is a commitment to backwards compatibility. I'd rather retain flexibility to break the internal implementation as needed.